docs: Navigator2.0: Design Document (#898)
platform-platform/monorepo: metrics/web/docs/features/navigation/01_navigation_design.md

Using the MetricsPage’s data directly in the UI(in some widget to be precise) or add logic to the widget itself will break the principle of separation between the presentation layer components and makes it hard to test the application logic.

Implementing the data (ViewModels) flow between pages/widgets using the result param of the pop method will break the principle of separation between the presentation layer components and makes it hard to test the application logic.

That's why using the MetricsPage's values is out of Metrics Web Architecture.

### Metrics Page Factory

The `MetricsPagesFactory` is a class that is responsible for creating a new [MetricsPage](#metrics-page) by the given `RouteConfiguration`. If the given `RouteConfiguration` matches with one of the predefined configurations, the factory creates a new `MetricsPage` to add it then to the current pages.

### Metrics Page

> The `MetricsPage` is an app-specific class that describes a configuration of the `Route` and uses in the `Navigator` page's argument. It configured to disable transition animation when replacing the entire screen.

Reviewer: ```suggestion
The `MetricsPage` is an app-specific implementation for the `Page` class that holds a configuration for the `Route`. The `MetricsPage` creates the `MetricsPageRoute` that disables transition animation on route changes.

The `NavigationNotifier` holds and manages the list of current `MetricsPage`s to use them in the `Navigator` constructor within the `MetricsRouterDelegate`.
```

### Route Configuration Factory

The `RouteConfigurationFactory` is responsible for creating the [RouteConfiguration](#route-information) from the given `URI`. In fact, it uses the `MetricsRoutes` class that holds a list of pre-defined `RouteConfiguration`s and maps the incoming URI to the `RouteConfiguration` related to the specific application page that matches the given `URI`.

The following diagram describes the structure and relationship between the above classes:

![Route change diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/navigation2_design_docs/metrics/web/docs/features/navigation/diagrams/navigation_class_diagram.puml)

## Making things work

The following section provides implementation details of a new navigation system for the Metrics Web Application. As navigation is related to the UI of the application, the required changes affect only the presentation layer. Read more about layers and their responsibilities in the [Metrics Web Application architecture document](https://github.com/platform-platform/monorepo/blob/master/metrics/web/docs/01_metrics_web_application_architecture.md).

> According to the above class diagram, we should implement several classes to integrate the Router class into the application. Here is a list of them providing short descriptions for each class: 

Reviewer: ```suggestion
According to the above class diagram, we should implement several classes to integrate the Router class into the application. Here is a list of them providing a short description for each class: 
```

- `MetricsRouteInformationParser` as the app-specific [`RouteInformationParser`](#route-information-parser) with the `RouteConfigurationFactory` to simplify parsing.
- `MetricsRouterDelegate`as the app-specific [`RouterDelegate`](#router-delegate).
- `NavigationNotifier` to manage pages and rebuild the `Navigator`.

> - `MetricsPageFactory` to simplify creating pages within the `NavigationNotifier`.
> Once the required classes are implemented and ready to use, we can migrate the application to the new navigation. In the `MetricsApp` we should replace the `MaterialApp` constructor with the `MaterialApp.router()` and inject the required fields. Follow the design examined in the class diagram to integrate all implemented classes. After all classes in place and configured, the new navigation is integrated!

Reviewer: ```suggestion
- `MetricsPageFactory` to simplify creating pages within the `NavigationNotifier`.

Once the required classes are implemented and ready to use, we can migrate the application to the new navigation. In the `MetricsApp` we should replace the `MaterialApp` constructor with the `MaterialApp.router()` and inject the required fields. Follow the design examined in the class diagram to integrate all implemented classes. After all classes are in place and configured, the new navigation is integrated!
```

### Route Configuration Factory

The `RouteConfigurationFactory` is responsible for creating the [RouteConfiguration](#route-information) from the given `URI`.

In fact, it uses the `MetricsAppRoutes` class that holds a list of pre-defined `RouteConfiguration`s and maps the incoming URI to the `RouteConfiguration` related to the specific application page.

The following diagram describes the structure and relationship between the above classes:

![Route change diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/navigation2_design_docs/metrics/web/docs/features/navigation/diagrams/navigation_class_diagram.puml)

## Making things work

The following section provides an implementation details of a new navigation integration into the Metrics Web Application. As navigation is related to the UI of the application, the required changes affect only the presentation layer. Read more about layers and their responsibilities in the [Metrics Web Application architecture document](https://github.com/platform-platform/monorepo/blob/master/metrics/web/docs/01_metrics_web_application_architecture.md). 

> 

Reviewer: ```suggestion

According to the above class diagram, we should implement several classes to integrate the Router class into the application. Here is a list of them providing short descriptions for each class: 
- `MetricsRouteInformationParser` as the app-specific [`RouteInformationParser`](#route-information-parser) with the `RouteConfigurationFactory` to simplify parsing.
- `MetricsRouterDelegate`as the app-specific [`RouterDelegate`](#router-delegate).
- `NavigationNotifier` to manage pages and rebuild the `Navigator`.
- `MetricsPageFactory` to simplify creating pages within the `NavigationNotifier`.

Once the required classes are implemented and ready to use, we can migrate the application to the new navigation. In the `MetricsApp` we should replace the `MaterialApp` constructor with the `MaterialApp.router()` and inject the required fields. Follow the design examined in the class diagram to integrate all implemented classes. After all classes in place and configured, the new navigation is integrated!

```

> To introduce this feature, we should follow the next steps:
> 
> 1. Replace the `MaterialApp` widget in the root `main.dart` file with the new `MaterialApp.router()` constructor.
> 2. Create the `MetricsRouteInformationParser` and pass it to the `MaterialApp.router()`, as it is the first required parameter.
> 3. Create the `RouteConfigurationFactory` class, which helps in parsing and mapping the `RouteInformation` in the created `MetricsRouteInformationParser`.
> 4. To help with mapping an incoming `RouteInformation` to the application-specific, we should define the `MetricsRoutes` class, which consists of pre-defined `RouteConfiguration`s that belongs to Metrics application pages.
> 5. Create the `MetricsRouterDelegate` and pass it to the `MaterialApp.router()`, as it is the second required parameter.
> 6. Create the `NavigationNotifier` and inject it to the application via `InjectorContainer`. Also, we should connect it with the `AuthNotifier` to receive the authentication updates once the `AuthNotifier` changes.
> 7. Provide methods that should have similar to the `Navigation 1.0` API, to make navigation in the application, such as: 
>      - pushNamed()
>      - popNamed()
>      - pushReplacement()
>      - pushReplacementNamed()
>      - pushAndRemoveWhere()
> 8. Integrate the `NavigationNotifier` into the `AppRouterDelegate` as well, to extract methods for the actual navigation and a list of pages, specific to the Metrics app from the `AppRouterDelegate`.
> 9. In the `AppRouterDelegate` provide the `setInitialRoutePath` and the `setNewRoutePath` methods to handle the navigation and `build` method to build the `Navigator` widget with the configured list of pages. Also, override the `currentConfiguration` getter to help the `Router` in updating the route information.
> 10. As the `MaterialApp.router` constructor does not contain the `observers` field, we should pass the `List<NavigatorObserver>` to the `MetricsRouterDelegate` class, which passes it to the `Navigator` widget. The existing observers remain unchanged. 
> 
> With that in place, we can use the provided methods from the `NavigationNotifier` to make navigation with the new navigation system.

Reviewer: ```suggestion
```

> The following sequence diagrams display the process of navigation with the new `Navigation 2.0`.

Reviewer: ```suggestion
The following sequence diagrams describe the navigation process using the new `Navigation 2.0` integrated.
```
